finality-aleph/src/new_network/split.rs

```diff
@@ -0,0 +1,231 @@
+        loop {
+            tokio::select! {
+                data = self.translated_receiver.next() => return data,
+                _ = forward_or_wait(&self.receiver, &self.left_sender, &self.right_sender) => (),
+            }
+        }
+    }
+}
+
+struct LeftNetwork<
+    D1: Data,
+    D2: Data,
+    S: SenderComponent<Split<D1, D2>>,
+    R: ReceiverComponent<Split<D1, D2>>,
+> {
+    sender: LeftSender<D1, D2, S>,
+    receiver: Arc<Mutex<LeftReceiver<D1, D2, R>>>,
+}
+
+impl<
+        D1: Data,
+        D2: Data,
+        S: SenderComponent<Split<D1, D2>>,
+        R: ReceiverComponent<Split<D1, D2>>,
+    > ComponentNetwork<D1> for LeftNetwork<D1, D2, S, R>
+{
+    type S = LeftSender<D1, D2, S>;
+    type R = LeftReceiver<D1, D2, R>;
+    fn sender(&self) -> &Self::S {
+        &self.sender
+    }
+    fn receiver(&self) -> Arc<Mutex<Self::R>> {
+        self.receiver.clone()
+    }
+}
+
+struct RightNetwork<
+    D1: Data,
+    D2: Data,
+    S: SenderComponent<Split<D1, D2>>,
+    R: ReceiverComponent<Split<D1, D2>>,
+> {
+    sender: RightSender<D1, D2, S>,
+    receiver: Arc<Mutex<RightReceiver<D1, D2, R>>>,
+}
+
+impl<
+        D1: Data,
+        D2: Data,
+        S: SenderComponent<Split<D1, D2>>,
+        R: ReceiverComponent<Split<D1, D2>>,
+    > ComponentNetwork<D2> for RightNetwork<D1, D2, S, R>
+{
+    type S = RightSender<D1, D2, S>;
+    type R = RightReceiver<D1, D2, R>;
+    fn sender(&self) -> &Self::S {
+        &self.sender
+    }
+    fn receiver(&self) -> Arc<Mutex<Self::R>> {
+        self.receiver.clone()
+    }
+}
+
+fn split_sender<D1: Data, D2: Data, S: SenderComponent<Split<D1, D2>>>(
+    sender: &S,
+) -> (LeftSender<D1, D2, S>, RightSender<D1, D2, S>) {
+    (
+        LeftSender {
+            sender: sender.clone(),
+            phantom: PhantomData,
+        },
+        RightSender {
+            sender: sender.clone(),
+            phantom: PhantomData,
+        },
+    )
+}
+
+fn split_receiver<D1: Data, D2: Data, R: ReceiverComponent<Split<D1, D2>>>(
+    receiver: Arc<Mutex<R>>,
+) -> (LeftReceiver<D1, D2, R>, RightReceiver<D1, D2, R>) {
+    let (left_sender, left_receiver) = mpsc::unbounded();
+    let (right_sender, right_receiver) = mpsc::unbounded();
+    (
+        LeftReceiver {
+            receiver: receiver.clone(),
+            translated_receiver: left_receiver,
+            left_sender: left_sender.clone(),
+            right_sender: right_sender.clone(),
+        },
+        RightReceiver {
+            receiver,
+            translated_receiver: right_receiver,
+            left_sender,
+            right_sender,
+        },
+    )
+}
+
+/// Split a single component network into two separate ones. This way multiple components can send
+/// data to the same underlying session not knowing what types of data the other ones use.
+///
+/// The main example for now is creating an `aleph_bft::Network` and a separate one for accumulating
+/// signatures for justifications.
```

Comment:
I would add a section describing how communication works inside.

Something like this (but worded a bit better): 
* Left and Right networks compete for the data produced by their parent network.
* Left can only handle LData type (Similarly Right can handle only RData)
* They can't know in advance what kind of data the sender from the parent network will send so they have to consume it anyway
* But they are actually nice and whenever they get data for the other network they are kind enough to forward it to their counterpart

Reply:
Added a description based on this.

Reply:
Thanks